src/devtools/store.js

```diff
@@ -870,6 +889,11 @@ export default class Store extends EventEmitter {
               throw new Error(`Node ${id} was removed before its children.`);
             }
 
+            // The following call depends on `getElementByID`
+            // which depends on the element being in `_idToElement`,
+            // so we have to do it before removing the element from `_idToElement`.
+            this._clearProfilingSnapshotRecursive(id);
```

Comment:
In a meeting so I'm a bit distracted, but I think this change would break things.

It's okay for us to clear _root_ data, because if you can't select the root (in the Components tree) then there's no way to view the Profiler charts to begin with, but it's _not_ safe to clear the Element data, because we need these initial snapshots to render the charts in the flame graph for elements that existed during some commits, but not during other commits.

So basically, I think this should be moved down into the `if (parentID === 0) {` block below (so that it's only done for roots).

Reply:
> It's okay for us to clear root data, because if you can't select the root (in the Components tree) then there's no way to view the Profiler charts to begin with, but it's not safe to clear the Element data, because we need these initial snapshots to render the charts in the flame graph for elements that existed during some commits, but not during other commits.

Sure, that's why we review code, I thought about this, too, and it was under the "is root?" condition initially, I just thought we might be leaking something if we don't. But now I think you're right, they should be cleaned up when their root cleans up recursively.

Reply:
So this _would_ have caused a runtime error, bug I think fixing it is a bit trickier than I assumed because of the fact that we unmount children before their parents. (This means that although the only time it's safe to clear this map is when a root is unmounted, it's also _too late_ to recursively clear it at the point.) 😁

Need to give this a little more thought.

Reply:
Looks like another source of complication is when we remove a root, the renderer interface's `flushPendingEvents()` gets a new ID for that root (since we have already cleared out the `fiberToIDMap`) and this causes the `Store` to add a new entry to the `_profilingOperationsByRootID` map 😁

Reply:
Awesome, thanks for looking, looks like this surfaced deeper problems with the implementation.

Reply:
I don't know if I'd categorize it that way 😁 we just weren't trying to reset on unmount before

Reply:
Yeah yeah YAGNI approach. I hate it, if you ask me, and prefer to think several steps forward at least in such memory/resource management scenarios where there's some alloc or add something, I believe there must be mirroring dealloc or remove, most of the times, otherwise you're shooting in the foot.

Reply:
I think it would be silly to suggest that either of us aren't thinking a few steps ahead 😄 We're just prioritizing differently.

Reply:
Agree. 😄 I do prioritize correct resource management as I've seen numerous examples of non-linear increase in bugs and effort to maintain the code where this is not addressed timely, if not built with this requirement initially.